compiler_opt/rl/imitation_learning/weighted_bc_trainer.py

```diff
@@ -0,0 +1,535 @@
+# coding=utf-8
+# Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Module for training an inlining policy with imitation learning."""
+
+from absl import app
+from absl import flags
+
+from typing import List, Optional
+
+import bisect
+import copy
+import gin
+import json
+import logging
+import os
+from functools import partial
+
+from compiler_opt.rl import policy_saver
+from compiler_opt.rl.imitation_learning.generate_bc_trajectories_lib import ProfilingDictValueType
+from compiler_opt.rl.imitation_learning.generate_bc_trajectories_lib import SequenceExampleFeatureNames
+from compiler_opt.rl.inlining import imitation_learning_config as config
+from compiler_opt.rl import feature_ops
+
+import keras
+import numpy as np
+import tensorflow as tf
+
+import tf_agents
+from tf_agents.trajectories import time_step as ts
+from tf_agents.typing import types
+from tf_agents.trajectories import policy_step
+import tensorflow_probability as tfp
+
+_QUANTILE_MAP_PATH = flags.DEFINE_string(
+    'quantile_map_path', None,
+    ('Directory containing the quantile map for normalizing features'
+     'in feature_ops.build_quantile_map.'))
+_TRAINING_DATA = flags.DEFINE_multi_string(
+    'training_data', None, 'Training data for one step of BC-Max')
+_PROFILING_DATA = flags.DEFINE_multi_string(
+    'profiling_data', None,
+    ('Paths to profile files for computing the TrainingWeights'
+     'If specified the order for each pair of json files is'
+     'comparator.json followed by eval.json and the number of'
+     'files should always be even.'))
+_SAVE_MODEL_DIR = flags.DEFINE_string(
+    'save_model_dir', None, 'Location to save the keras and TFAgents policies.')
+# _GIN_FILES = flags.DEFINE_multi_string(
+#     'gin_files', [], 'List of paths to gin configuration files.')
+# _GIN_BINDINGS = flags.DEFINE_multi_string(
+#     'gin_bindings', [],
+#     'Gin bindings to override the values set in the config files.')
+
+FLAGS = flags.FLAGS
+
+# Pytype cannot pick up the pyi file for tensorflow.summary. Disable the error
+# here as these errors are false positives.
+# pytype: disable=pyi-error
+
+
+@gin.configurable
+class TrainingWeights:
+  """Class for computing weights for training."""
+
```

Comment:
maybe add a brief comment on how one would use it, e.g. create an instance, then what... call create_new_profile... something something... repeat?